src/shared/tpm2-util.h

```diff
@@ -259,6 +269,173 @@ int tpm2_tpm2b_public_to_fingerprint(const TPM2B_PUBLIC *public, void **ret_fing
+                void **UNIQ_T(BUF, u) = (void**) &(buf);                \
+                size_t UNIQ_T(O, u) = (offset);                         \
+                size_t UNIQ_T(S, u);                                    \
+                _tpm2_marshal_size(UNIQ_T(DESC, u), UNIQ_T(SRC, u), &UNIQ_T(S, u), CONCATENATE(u, _tpm2_marshal_realloc)) \
+                        ? /* != 0 (failure) */                          \
+                        : greedy_realloc(UNIQ_T(BUF, u), UNIQ_T(O, u) + UNIQ_T(S, u), 1) \
+                        ? _tpm2_marshal(UNIQ_T(DESC, u), UNIQ_T(SRC, u), *UNIQ_T(BUF, u), UNIQ_T(O, u) + UNIQ_T(S, u), UNIQ_T(O, u), (ret_size), CONCATENATE(u, _tpm2_marshal_realloc)) \
+                        : log_oom();                                    \
+        })
+
+/* Unmarshal data from buf, starting at offset, into dst. The size of buf is max. If successful and ret_size
+ * is not NULL, it is set to offset plus the number of unmarshalled bytes. Returns 0 on success or < 0 on
+ * error. */
+#define tpm2_unmarshal(desc, dst, buf, max, offset, ret_size)           \
+        (tpm2_dlopen() ?: _tpm2_unmarshal(desc, dst, buf, max, offset, ret_size, UNIQ))
+#define _tpm2_unmarshal(desc, dst, buf, max, offset, ret_size, u)       \
+        ({                                                              \
+                const char *UNIQ_T(DESC, u) = (desc);                   \
+                log_debug("Unmarshalling %s", UNIQ_T(DESC, u));         \
+                size_t UNIQ_T(O, u) = (offset);                         \
+                TSS2_RC UNIQ_T(RC, u) = _UNMARSHAL(dst)(buf, max, &UNIQ_T(O, u), dst); \
+                UNIQ_T(RC, u) == TSS2_RC_SUCCESS     \
+                        ? _tpm2_marshalling_update_ret_size(UNIQ_T(O, u), (ret_size), CONCATENATE(u, _tpm2_unmarshal)) \
+                        : log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE), \
+                                          "Failed to unmarshal %s: %s", \
+                                          desc, sym_Tss2_RC_Decode(UNIQ_T(RC, u))); \
+        })
+
+#define tpm2_unmarshal_from_file(desc, dst, f, ret_size)                \
+        (tpm2_dlopen() ?: _tpm2_unmarshal_from_file(desc, dst, f, ret_size, UNIQ))
+#define _tpm2_unmarshal_from_file(desc, dst, f, ret_size, u)            \
+        ({                                                              \
+                _cleanup_free_ char *UNIQ_T(B, u) = NULL;               \
+                size_t UNIQ_T(S, u);                                    \
+                int UNIQ_T(R, u) = read_full_file(f, &UNIQ_T(B, u), &UNIQ_T(S, u)); \
+                if (UNIQ_T(R, u) >= 0)                                  \
+                        UNIQ_T(R, u) = _tpm2_unmarshal(desc, dst, (uint8_t*) UNIQ_T(B, u), UNIQ_T(S, u), 0, ret_size, CONCATENATE(u, _tpm2_unmarshal_from_file)); \
+                UNIQ_T(R, u);                                           \
```

Comment:
io errors will not be logged here, but marshalling errors will be.... we usually want that either all errors are logged or none